FACTORY-2997: implementation of get_source_components_for_build

twaugh opened the pull request:
Signed-off-by: Tim Waugh <twaugh@redhat.com>

lslebodn commented on tests/client/test_query.py:
```diff
@@ -55,3 +56,65 @@ def test_get_container_sources():
+    glibc_internal_source.upstream.connect(glibc_upstream_source)
+    glibc_component = Component(canonical_type='generic', canonical_namespace='foo',
+                                canonical_name='bar')
+    glibc_upstream_source.component.connect(glibc_component)
```
The test failed here:
```
self = <neomodel.cardinality.ZeroOrOne object at 0x7fbcdad979e8>
obj = <Component: {'canonical_namespace': 'foo', 'canonical_name': 'bar', 'canonical_type': 'generic', 'alternative_names': None}>
    def _check_node(self, obj):
        """check for valid node i.e correct class and is saved"""
        if not isinstance(obj, self.definition['node_class']):
            raise ValueError("Expected node of class " + self.definition['node_class'].__name__)
        if not hasattr(obj, 'id'):
>           raise ValueError("Can't perform operation on unsaved node " + repr(obj))
E           ValueError: Can't perform operation on unsaved node <Component: {'canonical_namespace': 'foo', 'canonical_name': 'bar', 'canonical_type': 'generic', 'alternative_names': None}>
/usr/lib/python3.6/site-packages/neomodel/relationship_manager.py:56: ValueError
```

lslebodn commented on tests/client/test_query.py:
```diff
@@ -55,3 +56,65 @@ def test_get_container_sources():
+
+    # The container itself
+    container_filename = ('docker-image-sha256:' + '1' * 64 + '.x86_64.tar.gz')
+    a = Artifact(
```
I would appreciate a little bit longer name then 1 character.

lslebodn commented on tests/client/test_query.py:
```diff
@@ -55,3 +56,65 @@ def test_get_container_sources():
+    # glibc RPM for parent image
+    glibc_build = Build(id_='770188', type_='rpm').save()
+    glibc_rpm = Artifact(
+        archive_id='5818103', architecture='x86_64', filename='glibc-1.2.3-1.el7.x86_64.rpm',
```
Is there any reason why we cannot have real NVR for glibc and real build id?

lslebodn replied:
BTW current version in centos 7 is `glibc-2.17-222.el7`+ buildID=651858

lslebodn commented on tests/client/test_query.py:
```diff
@@ -55,3 +56,65 @@ def test_get_container_sources():
+
+    # Parent image, p
+    container_filename = ('docker-image-sha256:' + '0' * 64 + '.x86_64.tar.gz')
+    p = Artifact(
```
I would appreciate a little bit longer name then 1 character.
It is simper to read assertions in following lines,

mprpic replied:
And doesn't break `(i)pdb` where `p` is the shorthand for the print expression.

twaugh replied:
How does it break pdb? Anyway, will make them longer. I think writing Go has made me appreciate shorter variable names :)

mprpic replied:
And doesn't break `(i)pdb` where `p` is the shorthand for the print expression.

twaugh replied:
I wasn't able to break pdb this way.

mprpic replied:
(Oops, sorry for the double comment, I didn't realize it submitted it the first time.)

```
$ cat pdb_test.py 
from ipdb import set_trace

p = 'hello'
parent = 'world'
set_trace()
$ python pdb_test.py 
3.6.6 (default, Jul 19 2018, 14:25:17)
--Return--
None
> ./pdb_test.py(5)<module>()
      3 p = 'hello'
      4 parent = 'world'
----> 5 set_trace()

ipdb> parent
'world'
ipdb> p
*** SyntaxError: unexpected EOF while parsing
ipdb> p p
'hello'
```

lslebodn commented on tests/client/test_query.py:
```diff
@@ -55,3 +56,65 @@ def test_get_container_sources():
+    p.embedded_artifacts.connect(glibc_rpm)
+
+    expected = []
+    assert query.get_source_components_in_archive('742663') == expected
```
Could you add comment why do we expect empty list here?

twaugh replied:
This was just work-in-progress data. Fixed now.

lslebodn reviewed (changes requested):
I added few comments

lslebodn commented:
@twaugh, Is that PR still WIP due to failure in the test *test_get_source_components_for_build*?

twaugh commented:
Please take another look.

twaugh commented:
Note: it currently does not return the result in a tree but has references to other elements. I'm working on a fix for that now.
UPDATE: fixed now.

mprahl commented on assayist/client/query.py:
```diff
@@ -44,3 +45,134 @@ def get_container_content_sources(build_id):
+    query = """
+    // Find artifacts which (or artifacts which embed artifacts which)...
+    // Note: "*0.." means zero or more edges; if zero edges, 'a' is the
+    //
```
Optional: Could you remove this line?

twaugh replied:
It was in the wrong place. Fixed.

mprahl commented on assayist/client/query.py:
```diff
@@ -44,3 +45,134 @@ def get_container_content_sources(build_id):
+
+    query = """
+    // Find the artifacts
+    // TODO: archive_id is not unique. We should take the
```
Are you planning to address this before merging the PR?

mprahl replied:
After thinking about this, you should consider storing the "internal" Neo4j ID of the artifact nodes queried above and just query directly by the ID here. This should be very fast and accurate.

Here is an example:
https://neo4j.com/docs/developer-manual/current/cypher/clauses/match/#match-multiple-nodes-by-id

twaugh replied:
Nice, thanks. Fixed.

mprahl commented on assayist/client/query.py:
```diff
@@ -44,3 +45,134 @@ def get_container_content_sources(build_id):
+        )
+
+        metadata = artifacts_by_id[(artifact.type_, artifact.archive_id)]
+        sources = metadata.setdefault("sources", [])
```
Optional: We use single quotes in the rest of the codebase, it'd be nice to be consistent to not have to switch back and forth when modifying existing code.

twaugh replied:
Think I got them all.

mprahl commented on assayist/client/query.py:
```diff
@@ -44,3 +45,134 @@ def get_container_content_sources(build_id):
+    results, _ = neomodel.db.cypher_query(query)
+
+    for a, sl, c in results:
+        artifact, sourceloc, component = (
```
Optional: What's the benefit of using this tuple? Seems like it'd take less lines and be easier to read to do the following:
```python
artifact = Artifact.inflate(a)
sourceloc = SourceLocation.inflate(sl)
component = Component.inflate(c)
```

mprahl commented on assayist/client/query.py:
```diff
@@ -44,3 +45,134 @@ def get_container_content_sources(build_id):
+    def construct(aid):
+        a = artifacts_by_id[aid]
+        try:
+            embeds_ids = a.pop("embeds_ids")
```
Why are you using `pop` instead of `a['embeds_ids']`? Is this so that you aren't stuck in an infinite loop in the event that an embedded artifact embeds the parent? I wouldn't think this could ever occur though.

twaugh replied:
I'm removing the key 'embeds_id' and replacing it with a new key 'embeds', whose value is a different type.

The 'embeds_id' key is not intended to be part of the final result.

mprahl commented on tests/client/test_query.py:
```diff
@@ -61,3 +62,198 @@ def test_get_container_sources():
+            'artifact': Artifact.nodes.get(archive_id='4001'),
+            'embeds': {
+                ('container', '2001'): {
+                    "artifact": Artifact.nodes.get(archive_id='2001'),
```
Why are you returning an `Artifact` object as part of your API function? I thought the API was supposed to return data that could be serialized to JSON.

twaugh replied:
Good point. Maybe the "artifact" key should be replaced by the keys and values of the node's properties? e.g.
```
'embeds': {
  ('container', '2001'): {
    'filename': ...,
    'architecture': ...,
    'sources': ...
    'embeds': ...
  }
}
```

Or is it better to have the properties in their own dict, separated from 'sources' and 'embeds' (which we're adding from other nodes and relationships)?

mprahl replied:
@twaugh could you show examples of what you mean?

twaugh replied:
Option 1:
```
expected = {
  ('container', '4001'): {
    'architecture': 'x86_64',
    'filename': 'docker-image-sha256:'  '1' * 64 + '.x86_64.tar.gz',
    'sources': [
      {
        'name': 'configmap-reload',
        'type': 'golang',
        ...
      },
      ...
    ],
    'embeds': {
      ('container', '2001'): {
        'architecture': 'x86_64',
        'filename': 'docker-image-sha256:' + '0' * 64 + '.x86_64.tar.gz',
        'embeds': { ... },
        ...
      },
      ...
    }
  },
  ...
}
```

Option 2:
```
expected = {
  ('container', '4001'): {
    'artifact': {
      'architecture': 'x86_64',
      'filename': 'docker-image-sha256:'  '1' * 64 + '.x86_64.tar.gz'
    },
    'sources': [
      {
        'name': 'configmap-reload',
        'type': 'golang',
        ...
      },
      ...
    ],
    'embeds': {
      ('container', '2001'): {
        'artifact': {
          'architecture': 'x86_64',
          'filename': 'docker-image-sha256:' + '0' * 64 + '.x86_64.tar.gz'
        },
        ...
      },
      ...
    }
  },
  ...
}
```

twaugh replied:
Fixed using option 2.

mprahl commented on tests/client/test_query.py:
```diff
@@ -61,3 +62,198 @@ def test_get_container_sources():
+
+def test_get_source_components_for_build():
+    """Test the get_source_components_for_build function."""
+    # Set up a layered container image build, build, whose single
```
What do you mean by `container image build, build`?

twaugh replied:
Reworded.

mprahl reviewed (changes requested):
I left some comments. Please let me know once they are addressed. Thank you for making this rather complex query and test data quite readable.

mprpic commented on assayist/client/query.py:
```diff
@@ -44,3 +45,134 @@ def get_container_content_sources(build_id):
+    artifacts.
+
+    :param int build_id: the Koji build's ID
+    :return: a list of canonical names (package URLs)
```
I guess this should be updated to `a tree of artifacts and their source code components`?

twaugh replied:
Thanks, fixed.

sdherr reviewed (approved):
Makes sense to me.

mprahl commented:
@twaugh everything looks good except that it looks like Artifact objects are still returned. Could you please serialize those as we discussed above?

twaugh commented:
Which option should I use? See https://github.com/release-engineering/assayist/pull/13#discussion_r225211785

mprahl commented:
> Which option should I use? See [#13 (comment)](https://github.com/release-engineering/assayist/pull/13#discussion_r225211785)

I'm sorry for not seeing that. I personally prefer option #2 because it keeps the nodes separate and it'll be easier to keep that formatting across the other APIs.

twaugh committed:
FACTORY-2997: implementation of get_source_components_for_build

Signed-off-by: Tim Waugh <twaugh@redhat.com>

twaugh committed:
get_source_components_for_build: handle multiple SOURCE_FOR relationships

A Component node may be related to multiple SourceLocation nodes. To
handle this, the query needs to avoid using SourceLocations that have a
further upstream SourceLocation when gathering fields for the canonical
name.

Signed-off-by: Tim Waugh <twaugh@redhat.com>

twaugh commented:
Updated. Please take another look.

I added another commit to link in Component relationships to all SourceLocations, not just the most-upstream one. This also needed a change to the query (in the same commit).

mprahl reviewed (approved):
:+1:
